pkg/platform/kube/provisioner/apigateway/provisioner.go

```diff
@@ -0,0 +1,369 @@
+	kubeClientSet   kubernetes.Interface
+	nuclioClientSet nuclioio_client.Interface
+	ingressManager  *ingress.Manager
+}
+
+func NewProvisioner(loggerInstance logger.Logger,
+	kubeClientSet kubernetes.Interface,
+	nuclioClientSet nuclioio_client.Interface,
+	ingressManager *ingress.Manager) (*Provisioner, error) {
+
+	newProvisioner := &Provisioner{
+		Logger:          loggerInstance.GetChild("apigateway"),
+		kubeClientSet:   kubeClientSet,
+		nuclioClientSet: nuclioClientSet,
+		ingressManager:  ingressManager,
+	}
+
+	return newProvisioner, nil
+}
+
+func (p *Provisioner) CreateOrUpdateAPIGateway(ctx context.Context, apiGateway *nuclioio.NuclioAPIGateway) error {
+	var appliedIngressNames []string
+
+	apiGateway.Status.Name = apiGateway.Spec.Name
+
+	if err := p.validateSpec(apiGateway); err != nil {
+		return errors.Wrap(err, "Api gateway spec validation failed")
+	}
+
+	// generate an ingress for each upstream
+	upstreams := apiGateway.Spec.Upstreams
+	ingresses := map[string]*ingress.Resources{}
+
+	// always try to remove previous canary ingress first, because
+	// nginx returns 503 on all requests if primary service == secondary service. (happens on every promotion)
+	// so during promotion all requests will be sent to the primary ingress
+	p.tryRemovePreviousCanaryIngress(ctx, apiGateway)
+
+	if len(upstreams) == 1 {
+
+		// create just a single ingress
+		ingressResources, err := p.generateNginxIngress(ctx, apiGateway, upstreams[0])
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate nginx ingress")
+		}
+
+		ingresses[ingressResources.Ingress.Name] = ingressResources
+
+	} else if len(upstreams) == 2 {
+		var canaryUpstream platform.APIGatewayUpstreamSpec
+		var baseUpstream platform.APIGatewayUpstreamSpec
+
+		// determine which upstream is the canary one
+		if upstreams[0].Percentage != 0 {
+			baseUpstream = upstreams[1]
+			canaryUpstream = upstreams[0]
+		} else if upstreams[1].Percentage != 0 {
+			baseUpstream = upstreams[0]
+			canaryUpstream = upstreams[1]
+		} else {
+			return errors.New("Percentage must be set on one of the upstreams (canary)")
+		}
+
+		// validity check
+		if canaryUpstream.Percentage > 100 || canaryUpstream.Percentage < 1 {
+			return errors.New("The canary upstream percentage must be between 1 and 100")
+		}
+
+		// add the base ingress
+		baseIngressResources, err := p.generateNginxIngress(ctx, apiGateway, baseUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the base nginx ingress")
+		}
+		ingresses[baseIngressResources.Ingress.Name] = baseIngressResources
+
+		// add the canary ingress
+		canaryIngressResources, err := p.generateNginxIngress(ctx, apiGateway, canaryUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the canary nginx ingress")
+		}
+		ingresses[canaryIngressResources.Ingress.Name] = canaryIngressResources
+	}
+
+	// create ingresses
+	for ingressName, ingressResources := range ingresses {
```

Comment:
can that happen in parallel? is order matter? if not, you can use `errorGrroup`

Reply:
order matters - primary must be created before the canary one, because having only canary breaks this endpoint (nginx behaviour)

Reply:
great, can you please state that in a comment?